discovery-provider/src/eth_indexing/event_scanner.py

```diff
@@ -0,0 +1,434 @@
+from src.utils.helpers import redis_set_and_dump, redis_get_or_restore
+from src.queries.get_balances import enqueue_balance_refresh
+
+
+logger = logging.getLogger(__name__)
+
+eth_indexing_last_scanned_block_key = "eth_indexing_last_scanned_block"
+
+# How many times we try to re-attempt a failed JSON-RPC call
+MAX_REQUEST_RETRIES = 30
+# Delay between failed requests to let JSON-RPC server to recover
+REQUEST_RETRY_SECONDS = 3
+# Minimum number of blocks to scan for our JSON-RPC throttling parameters
+MIN_SCAN_CHUNK_SIZE = 10
+# How many maximum blocks at the time we request from JSON-RPC
+# and we are unlikely to exceed the response size limit of the JSON-RPC server
+MAX_CHUNK_SCAN_SIZE = 10000
+# Factor how was we increase chunk size if no results found
+CHUNK_SIZE_INCREASE = 2
+# initial number of blocks to scan, this number will increase/decrease as a function of whether transfer events have been found within the range of blocks scanned
+START_CHUNK_SIZE = 20
+# how many blocks from tail of chain we want to scan to
+ETH_BLOCK_TAIL_OFFSET = 1
+# the block number to start with if first time scanning
+# this should be the first block during and after which $AUDIO transfer events started occurring
+MIN_SCAN_START_BLOCK = 0
+
+
+class TransferEvent(TypedDict):
+    logIndex: int
+    transactionHash: Any
+    blockNumber: int
+    args: Any
+
+
+class EventScanner:
+    """Scan blockchain for events and try not to abuse JSON-RPC API too much.
+
+    Can be used for real-time scans, as it detects minor chain reorganisation and rescans.
+    Unlike the easy web3.contract.Contract, this scanner can scan events from multiple contracts at once.
+    For example, you can get all transfers from all tokens in the same scan.
+
+    You *should* disable the default `http_retry_request_middleware` on your provider for Web3,
+    because it cannot correctly throttle and decrease the `eth_get_logs` block number range.
+    """
+
+    def __init__(
+        self,
+        db,
+        redis,
+        web3: Web3,
+        contract: Type[Contract],
+        event_type: Type[ContractEvent],
+        filters: dict,
+    ):
+        """
+        :param db: database handle
+        :param redis: redis handle
+        :param web3: Web3 instantiated with provider url
+        :param contract: Contract
+        :param state: state manager to keep tracks of last scanned block and persisting events to db
+        :param event_type: web3 Event we scan
+        :param filters: Filters passed to get_logs e.g. { "address": <token-address> }
+        """
+
+        self.logger = logger
+        self.db = db
+        self.redis = redis
+        self.contract = contract
+        self.web3 = web3
+        self.event_type = event_type
+        self.filters = filters
+        self.last_scanned_block = MIN_SCAN_START_BLOCK
+        # self.state = EventScannerState(db, redis)
+        # self.state.restore()
+
+    def restore(self):
```

Comment:
If we put it in Redis, I think that means we can't include this in a DB dump and each new DN will have to cold start. Maybe this could live as a single row in the DB?

Reply:
this is a really good point^ agreed. a single row table just like "alembic_version" would do this justice